tests/components/opnsense/test_device_tracker.py

```diff
@@ -0,0 +1,60 @@
+"""The tests for the opnsense device tracker platform."""
+
+from unittest import mock
+
+from homeassistant.components.opnsense import CONF_API_SECRET, DOMAIN
+from homeassistant.const import CONF_URL, CONF_API_KEY, CONF_VERIFY_SSL
+from homeassistant.setup import async_setup_component
+import homeassistant.components.device_tracker as device_tracker
+
+from tests.common import MockDependency
+
+
+async def test_get_scanner(hass):
+    """Test creating an opnsense scanner."""
+    with MockDependency("pyopnsense") as mocked_opnsense:
+        interface_client = mock.MagicMock()
+        mocked_opnsense.diagnostics.InterfaceClient.return_value = interface_client
+        interface_client.get_arp.return_value = [
+            {
+                "hostname": "",
+                "intf": "igb1",
+                "intf_description": "LAN",
+                "ip": "192.168.0.123",
+                "mac": "ff:ff:ff:ff:ff:ff",
+                "manufacturer": "",
+            },
+            {
+                "hostname": "Desktop",
+                "intf": "igb1",
+                "intf_description": "LAN",
+                "ip": "192.168.0.167",
+                "mac": "ff:ff:ff:ff:ff:fe",
+                "manufacturer": "OEM",
+            },
+        ]
+        network_insight_client = mock.MagicMock()
+        mocked_opnsense.diagnostics.NetworkInsightClient = network_insight_client
+        network_insight_client.get_interfaces.return_value = {
+            "igb0": "WAN",
+            "igb1": "LAN",
+        }
+
+        result = await async_setup_component(
+            hass,
+            device_tracker.DOMAIN,
```

Comment:
We should set up the opnsense component not the device tracker component. The opnsense component will load its device tracker platform.

Reply:
I had it this way originally (you can see earlier iterations) but found during local testing that this didn't happen until the test teardown. By making it manually load device_tracker here it happened during the test which is what we wanted. If you have a suggestion on how to make the test wait for the device tracker component to be setup during. Also when this is coupled to your above review the mock doesn't apply to all the http calls made in teardown and it leads to tracebacks from failing to connect

Reply:
With just this change (still with a runtime import):

```
------------------------------ Captured log call -------------------------------
INFO     homeassistant.loader:loader.py:212 Loaded test_package from custom_components.test_package
INFO     homeassistant.loader:loader.py:212 Loaded test from custom_components.test
INFO     homeassistant.loader:loader.py:212 Loaded opnsense from homeassistant.components.opnsense
INFO     homeassistant.setup:setup.py:145 Setting up opnsense
INFO     homeassistant.setup:setup.py:185 Setup of domain opnsense took 0.0 seconds.
DEBUG    homeassistant.core:core.py:575 Bus:Handling <Event component_loaded[L]: component=opnsense>
INFO     homeassistant.loader:loader.py:212 Loaded device_tracker from homeassistant.components.device_tracker
---------------------------- Captured log teardown -----------------------------
DEBUG    homeassistant.core:core.py:575 Bus:Handling <Event homeassistant_stop[L]>
INFO     homeassistant.loader:loader.py:212 Loaded group from homeassistant.components.group
INFO     homeassistant.loader:loader.py:212 Loaded zone from homeassistant.components.zone
INFO     homeassistant.setup:setup.py:145 Setting up group
DEBUG    homeassistant.core:core.py:575 Bus:Handling <Event service_registered[L]: domain=group, service=reload>
DEBUG    homeassistant.core:core.py:575 Bus:Handling <Event service_registered[L]: domain=group, service=set>
DEBUG    homeassistant.core:core.py:575 Bus:Handling <Event service_registered[L]: domain=group, service=remove>
DEBUG    homeassistant.core:core.py:575 Bus:Handling <Event service_registered[L]: domain=group, service=set_visibility>
INFO     homeassistant.setup:setup.py:185 Setup of domain group took 0.0 seconds.
DEBUG    homeassistant.core:core.py:575 Bus:Handling <Event component_loaded[L]: component=group>
INFO     homeassistant.setup:setup.py:145 Setting up zone
INFO     homeassistant.setup:setup.py:185 Setup of domain zone took 0.0 seconds.
DEBUG    homeassistant.core:core.py:575 Bus:Handling <Event component_loaded[L]: component=zone>
DEBUG    homeassistant.core:core.py:575 Bus:Handling <Event state_changed[L]: entity_id=zone.home, old_state=None, new_state=<state zone.home=zoning; hidden=True, latitude=32.87336, longitude=-117.22743, radius=100, friendly_name=test home, icon=mdi:home @ 2019-10-02T19:12:43.197361+00:00>>
INFO     homeassistant.setup:setup.py:145 Setting up device_tracker
DEBUG    homeassistant.core:core.py:575 Bus:Handling <Event service_registered[L]: domain=device_tracker, service=see>
DEBUG    homeassistant.core:core.py:575 Bus:Handling <Event call_service[L]: domain=group, service=set, service_data=object_id=all_devices, visible=False, name=all devices, entities=['device_tracker.ff_ff_ff_ff_ff_ff', 'device_tracker.desktop']>
DEBUG    homeassistant.helpers.restore_state:restore_state.py:84 Not creating cache - no saved states found
DEBUG    homeassistant.core:core.py:575 Bus:Handling <Event state_changed[L]: entity_id=device_tracker.ff_ff_ff_ff_ff_ff, old_state=None, new_state=<state device_tracker.ff_ff_ff_ff_ff_ff=not_home; source_type=None, friendly_name=ff ff ff ff ff ff @ 2019-10-02T19:12:43.208801+00:00>>
DEBUG    homeassistant.core:core.py:575 Bus:Handling <Event state_changed[L]: entity_id=device_tracker.desktop, old_state=None, new_state=<state device_tracker.desktop=not_home; source_type=None, friendly_name=Desktop @ 2019-10-02T19:12:43.209131+00:00>>
INFO     homeassistant.setup:setup.py:185 Setup of domain device_tracker took 0.0 seconds.
DEBUG    homeassistant.core:core.py:575 Bus:Handling <Event component_loaded[L]: component=device_tracker>
DEBUG    homeassistant.core:core.py:575 Bus:Handling <Event platform_discovered[L]: service=load_platform.device_tracker, platform=opnsense, discovered=[]>
DEBUG    homeassistant.core:core.py:575 Bus:Handling <Event state_changed[L]: entity_id=group.all_devices, old_state=None, new_state=<state group.all_devices=not_home; entity_id=('device_tracker.ff_ff_ff_ff_ff_ff', 'device_tracker.desktop'), order=0, auto=True, friendly_name=all devices, hidden=True @ 2019-10-02T19:12:43.210215+00:00>>
INFO     homeassistant.components.device_tracker:setup.py:55 Setting up device_tracker.legacy
DEBUG    homeassistant.core:core.py:575 Bus:Handling <Event state_changed[L]: entity_id=device_tracker.ff_ff_ff_ff_ff_ff, old_state=<state device_tracker.ff_ff_ff_ff_ff_ff=not_home; source_type=None, friendly_name=ff ff ff ff ff ff @ 2019-10-02T19:12:43.208801+00:00>, new_state=<state device_tracker.ff_ff_ff_ff_ff_ff=home; source_type=router, latitude=32.87336, longitude=-117.22743, gps_accuracy=0, scanner=OPNSenseDeviceScanner, friendly_name=ff ff ff ff ff ff @ 2019-10-02T19:12:43.214202+00:00>>
DEBUG    homeassistant.core:core.py:575 Bus:Handling <Event state_changed[L]: entity_id=group.all_devices, old_state=<state group.all_devices=not_home; entity_id=('device_tracker.ff_ff_ff_ff_ff_ff', 'device_tracker.desktop'), order=0, auto=True, friendly_name=all devices, hidden=True @ 2019-10-02T19:12:43.210215+00:00>, new_state=<state group.all_devices=home; entity_id=('device_tracker.ff_ff_ff_ff_ff_ff', 'device_tracker.desktop'), order=0, auto=True, friendly_name=all devices, hidden=True @ 2019-10-02T19:12:43.215051+00:00>>
DEBUG    homeassistant.core:core.py:575 Bus:Handling <Event state_changed[L]: entity_id=device_tracker.desktop, old_state=<state device_tracker.desktop=not_home; source_type=None, friendly_name=Desktop @ 2019-10-02T19:12:43.209131+00:00>, new_state=<state device_tracker.desktop=home; source_type=router, latitude=32.87336, longitude=-117.22743, gps_accuracy=0, scanner=OPNSenseDeviceScanner, manufacturer=OEM, friendly_name=Desktop @ 2019-10-02T19:12:43.215551+00:00>>
DEBUG    homeassistant.core:core.py:575 Bus:Handling <Event homeassistant_close[L]>
```

Reply:
With this change and a module level import:

```
------------------------------ Captured log call -------------------------------
INFO     homeassistant.loader:loader.py:212 Loaded test_package from custom_components.test_package
INFO     homeassistant.loader:loader.py:212 Loaded test from custom_components.test
INFO     homeassistant.loader:loader.py:212 Loaded opnsense from homeassistant.components.opnsense
INFO     homeassistant.setup:setup.py:145 Setting up opnsense
INFO     homeassistant.setup:setup.py:185 Setup of domain opnsense took 0.0 seconds.
DEBUG    homeassistant.core:core.py:575 Bus:Handling <Event component_loaded[L]: component=opnsense>
---------------------------- Captured log teardown -----------------------------
DEBUG    homeassistant.core:core.py:575 Bus:Handling <Event homeassistant_stop[L]>
INFO     homeassistant.loader:loader.py:212 Loaded group from homeassistant.components.group
INFO     homeassistant.loader:loader.py:212 Loaded zone from homeassistant.components.zone
INFO     homeassistant.setup:setup.py:145 Setting up group
DEBUG    homeassistant.core:core.py:575 Bus:Handling <Event service_registered[L]: domain=group, service=reload>
DEBUG    homeassistant.core:core.py:575 Bus:Handling <Event service_registered[L]: domain=group, service=set>
DEBUG    homeassistant.core:core.py:575 Bus:Handling <Event service_registered[L]: domain=group, service=remove>
DEBUG    homeassistant.core:core.py:575 Bus:Handling <Event service_registered[L]: domain=group, service=set_visibility>
INFO     homeassistant.setup:setup.py:185 Setup of domain group took 0.0 seconds.
DEBUG    homeassistant.core:core.py:575 Bus:Handling <Event component_loaded[L]: component=group>
INFO     homeassistant.setup:setup.py:145 Setting up zone
INFO     homeassistant.setup:setup.py:185 Setup of domain zone took 0.0 seconds.
DEBUG    homeassistant.core:core.py:575 Bus:Handling <Event component_loaded[L]: component=zone>
DEBUG    homeassistant.core:core.py:575 Bus:Handling <Event state_changed[L]: entity_id=zone.home, old_state=None, new_state=<state zone.home=zoning; hidden=True, latitude=32.87336, longitude=-117.22743, radius=100, friendly_name=test home, icon=mdi:home @ 2019-10-02T19:23:57.596806+00:00>>
INFO     homeassistant.setup:setup.py:145 Setting up device_tracker
DEBUG    homeassistant.core:core.py:575 Bus:Handling <Event service_registered[L]: domain=device_tracker, service=see>
DEBUG    homeassistant.core:core.py:575 Bus:Handling <Event call_service[L]: domain=group, service=set, service_data=object_id=all_devices, visible=False, name=all devices, entities=['device_tracker.ff_ff_ff_ff_ff_ff', 'device_tracker.desktop']>
DEBUG    homeassistant.helpers.restore_state:restore_state.py:84 Not creating cache - no saved states found
DEBUG    homeassistant.core:core.py:575 Bus:Handling <Event state_changed[L]: entity_id=device_tracker.ff_ff_ff_ff_ff_ff, old_state=None, new_state=<state device_tracker.ff_ff_ff_ff_ff_ff=not_home; source_type=None, friendly_name=ff ff ff ff ff ff @ 2019-10-02T19:23:57.606378+00:00>>
DEBUG    homeassistant.core:core.py:575 Bus:Handling <Event state_changed[L]: entity_id=device_tracker.desktop, old_state=None, new_state=<state device_tracker.desktop=not_home; source_type=None, friendly_name=Desktop @ 2019-10-02T19:23:57.606559+00:00>>
INFO     homeassistant.setup:setup.py:185 Setup of domain device_tracker took 0.0 seconds.
DEBUG    homeassistant.core:core.py:575 Bus:Handling <Event component_loaded[L]: component=device_tracker>
DEBUG    homeassistant.core:core.py:575 Bus:Handling <Event platform_discovered[L]: service=load_platform.device_tracker, platform=opnsense, discovered=[]>
DEBUG    homeassistant.core:core.py:575 Bus:Handling <Event state_changed[L]: entity_id=group.all_devices, old_state=None, new_state=<state group.all_devices=not_home; entity_id=('device_tracker.ff_ff_ff_ff_ff_ff', 'device_tracker.desktop'), order=0, auto=True, friendly_name=all devices, hidden=True @ 2019-10-02T19:23:57.607193+00:00>>
INFO     homeassistant.components.device_tracker:setup.py:55 Setting up device_tracker.legacy
DEBUG    urllib3.connectionpool:connectionpool.py:959 Starting new HTTPS connection (1): fake_host_fun:443
ERROR    homeassistant.core:core.py:148 Error doing job: Task exception was never retrieved
Traceback (most recent call last):
  File "/home/mtreinish/git/home-assistant/.tox/py37/lib/python3.7/site-packages/urllib3/connection.py", line 157, in _new_conn
    (self._dns_host, self.port), self.timeout, **extra_kw
  File "/home/mtreinish/git/home-assistant/.tox/py37/lib/python3.7/site-packages/urllib3/util/connection.py", line 61, in create_connection
    for res in socket.getaddrinfo(host, port, family, socket.SOCK_STREAM):
  File "/usr/lib64/python3.7/socket.py", line 748, in getaddrinfo
    for res in _socket.getaddrinfo(host, port, family, type, proto, flags):
socket.gaierror: [Errno -2] Name or service not known

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/mtreinish/git/home-assistant/.tox/py37/lib/python3.7/site-packages/urllib3/connectionpool.py", line 672, in urlopen
    chunked=chunked,
  File "/home/mtreinish/git/home-assistant/.tox/py37/lib/python3.7/site-packages/urllib3/connectionpool.py", line 376, in _make_request
    self._validate_conn(conn)
  File "/home/mtreinish/git/home-assistant/.tox/py37/lib/python3.7/site-packages/urllib3/connectionpool.py", line 994, in _validate_conn
    conn.connect()
  File "/home/mtreinish/git/home-assistant/.tox/py37/lib/python3.7/site-packages/urllib3/connection.py", line 334, in connect
    conn = self._new_conn()
  File "/home/mtreinish/git/home-assistant/.tox/py37/lib/python3.7/site-packages/urllib3/connection.py", line 169, in _new_conn
    self, "Failed to establish a new connection: %s" % e
urllib3.exceptions.NewConnectionError: <urllib3.connection.VerifiedHTTPSConnection object at 0x7fb6b40b3ed0>: Failed to establish a new connection: [Errno -2] Name or service not known

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/mtreinish/git/home-assistant/.tox/py37/lib/python3.7/site-packages/requests/adapters.py", line 449, in send
    timeout=timeout
  File "/home/mtreinish/git/home-assistant/.tox/py37/lib/python3.7/site-packages/urllib3/connectionpool.py", line 720, in urlopen
    method, url, error=e, _pool=self, _stacktrace=sys.exc_info()[2]
  File "/home/mtreinish/git/home-assistant/.tox/py37/lib/python3.7/site-packages/urllib3/util/retry.py", line 436, in increment
    raise MaxRetryError(_pool, url, error or ResponseError(cause))
urllib3.exceptions.MaxRetryError: HTTPSConnectionPool(host='fake_host_fun', port=443): Max retries exceeded with url: /api/diagnostics/interface/getArp (Caused by NewConnectionError('<urllib3.connection.VerifiedHTTPSConnection object at 0x7fb6b40b3ed0>: Failed to establish a new connection: [Errno -2] Name or service not known'))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/mtreinish/git/home-assistant/homeassistant/components/device_tracker/setup.py", line 164, in async_device_tracker_scan
    found_devices = await scanner.async_scan_devices()
  File "/usr/lib64/python3.7/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/home/mtreinish/git/home-assistant/homeassistant/components/opnsense/device_tracker.py", line 39, in scan_devices
    self.update_info()
  File "/home/mtreinish/git/home-assistant/homeassistant/components/opnsense/device_tracker.py", line 55, in update_info
    devices = self.client.get_arp()
  File "/home/mtreinish/git/home-assistant/.tox/py37/lib/python3.7/site-packages/pyopnsense/diagnostics.py", line 52, in get_arp
    return self._get('diagnostics/interface/getArp')
  File "/home/mtreinish/git/home-assistant/.tox/py37/lib/python3.7/site-packages/pyopnsense/client.py", line 53, in _get
    timeout=DEFAULT_TIMEOUT)
  File "/home/mtreinish/git/home-assistant/.tox/py37/lib/python3.7/site-packages/requests/api.py", line 75, in get
    return request('get', url, params=params, **kwargs)
  File "/home/mtreinish/git/home-assistant/.tox/py37/lib/python3.7/site-packages/requests/api.py", line 60, in request
    return session.request(method=method, url=url, **kwargs)
  File "/home/mtreinish/git/home-assistant/.tox/py37/lib/python3.7/site-packages/requests/sessions.py", line 533, in request
    resp = self.send(prep, **send_kwargs)
  File "/home/mtreinish/git/home-assistant/.tox/py37/lib/python3.7/site-packages/requests/sessions.py", line 646, in send
    r = adapter.send(request, **kwargs)
  File "/home/mtreinish/git/home-assistant/.tox/py37/lib/python3.7/site-packages/requests/adapters.py", line 516, in send
    raise ConnectionError(e, request=request)
requests.exceptions.ConnectionError: HTTPSConnectionPool(host='fake_host_fun', port=443): Max retries exceeded with url: /api/diagnostics/interface/getArp (Caused by NewConnectionError('<urllib3.connection.VerifiedHTTPSConnection object at 0x7fb6b40b3ed0>: Failed to establish a new connection: [Errno -2] Name or service not known'))
DEBUG    homeassistant.core:core.py:575 Bus:Handling <Event homeassistant_close[L]>
```

Reply:
We need to `await hass.async_block_till_done()` after component setup to let the platform load.

Reply:
I would put the dependency mock etc in a pytest fixture function. That way it's applied during the whole test.

Reply:
Done in: https://github.com/home-assistant/home-assistant/pull/26834/commits/600cbd59ce571a9eb6b51ece5ef65e0361109280